substrate/primitives/core/src/pop.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities for proving possession of a particular public key
+
+use crate::crypto::{CryptoType, Pair};
+use sp_std::vec::Vec;
+
+/// Pair which is able to generate proof of possession.
+///
+/// This is implemented in different trait to provide default behavior.
+pub trait ProofOfPossessionGenerator: Pair
+where
+	Self::Public: CryptoType,
+{
+	/// Generate proof of possession.
+	///
+	/// The proof of possession generator is supposed to
+	/// produce a "signature" with unique hash context that should
+	/// never be used in other signatures. This proves that
+	/// the secret key is known to the prover. While prevent
+	/// malicious actors to trick an honest party to sign an
+	/// unpossessed public key resulting in a rogue key attack (See: Section 4.3 of
+	/// - Ristenpart, T., & Yilek, S. (2007). The power of proofs-of-possession: Securing multiparty
+	///   signatures against rogue-key attacks. In , Annual {{International Conference}} on the
+	///   {{Theory}} and {{Applications}} of {{Cryptographic Techniques} (pp. 228–245). : Springer.
+	#[cfg(feature = "full_crypto")]
+	fn generate_proof_of_possession(&mut self) -> Self::Signature;
+}
+
+/// Pair which is able to generate proof of possession.
+///
+/// While you don't need a keypair to verify a proof of possession (you only need a public key)
+/// we constrain on Pair to use the Public and Signature types associated to Pair.
+/// This is implemented in different trait (than Public Key) to provide default behavior.
```

Comment:
```suggestion
///
/// While you don't need a keypair to verify a proof of possession (you only need a public key)
/// we constrain on Pair to use the Public and Signature types associated to Pair.
/// This is implemented in different trait (than Public Key) to provide default behavior.
```

These are just code comments and clearly should not be docs. Also they are very confusing too read.